From bb8e57b57cd5dcaa78f6faf96a94d77c420e38cb Mon Sep 17 00:00:00 2001
From: Tellen Yu <tellen.yu@amlogic.com>
Date: Tue, 4 Jun 2019 18:03:06 +0800
Subject: [PATCH] Set bilinear filter quality when drawing VectorDrawable

PD#OTT-3699

Problem:
VectorDrawable can not be shown correctly

Solution:
This patch from Google:
    From 038fc3732d547d78227da2d1ef6898213e05b153 Mon Sep 17 00:00:00 2001
    From: Stan Iliev <stani@google.com>
    Date: Mon, 30 Jul 2018 18:31:46 -0400
    Subject: [PATCH] Set bilinear filter quality when drawing VectorDrawable

    Set filter quality to GL_LINEAR, when drawing
    VectorDrawable cache into the frame buffer.

    Bug: 77841258
    Test: Search bar icon jagged edge in launcher fixed
    Change-Id: I91270f64964ae3ba7a47264aec395bc2c2d9e968

Verify:
vefrify it on Franklin

Change-Id: I4a6fa86a4cc0909fdab5ac57b0ed31f07c031791
Signed-off-by: Tellen Yu <tellen.yu@amlogic.com>
---
 libs/hwui/VectorDrawable.cpp                    | 9 ++++-----
 libs/hwui/pipeline/skia/SkiaRecordingCanvas.cpp | 5 ++---
 2 files changed, 6 insertions(+), 8 deletions(-)

diff --git a/libs/hwui/VectorDrawable.cpp b/libs/hwui/VectorDrawable.cpp
index 18358e2..becfc22 100644
--- a/libs/hwui/VectorDrawable.cpp
+++ b/libs/hwui/VectorDrawable.cpp
@@ -474,14 +474,13 @@ SkPaint* Tree::getPaint() {
 // Update the given paint with alpha and color filter. Return nullptr if no color filter is
 // specified and root alpha is 1. Otherwise, return updated paint.
 SkPaint* Tree::updatePaint(SkPaint* outPaint, TreeProperties* prop) {
-    if (prop->getRootAlpha() == 1.0f && prop->getColorFilter() == nullptr) {
-        return nullptr;
-    } else {
+     // HWUI always draws VD with bilinear filtering.
+     outPaint->setFilterQuality(kLow_SkFilterQuality);
+     if (prop->getRootAlpha() < 1.0f || prop->getColorFilter() != nullptr) {
         outPaint->setColorFilter(sk_ref_sp(prop->getColorFilter()));
-        outPaint->setFilterQuality(kLow_SkFilterQuality);
         outPaint->setAlpha(prop->getRootAlpha() * 255);
-        return outPaint;
     }
+    return outPaint;
 }
 
 Bitmap& Tree::getBitmapUpdateIfDirty() {
diff --git a/libs/hwui/pipeline/skia/SkiaRecordingCanvas.cpp b/libs/hwui/pipeline/skia/SkiaRecordingCanvas.cpp
index f0da660..8eaeff2 100644
--- a/libs/hwui/pipeline/skia/SkiaRecordingCanvas.cpp
+++ b/libs/hwui/pipeline/skia/SkiaRecordingCanvas.cpp
@@ -220,9 +220,8 @@ void SkiaRecordingCanvas::drawNinePatch(Bitmap& bitmap, const Res_png_9patch& ch
     sk_sp<SkColorFilter> colorFilter;
     sk_sp<SkImage> image = bitmap.makeImage(&colorFilter);
     const SkPaint* filteredPaint = bitmapPaint(paint, &tmpPaint, colorFilter);
-    // Besides kNone, the other three SkFilterQualities are treated the same. And Android's
-    // Java API only supports kLow and kNone anyway.
-    if (!filteredPaint || filteredPaint->getFilterQuality() == kNone_SkFilterQuality) {
+    // HWUI always draws 9-patches with bilinear filtering, regardless of what is set in the Paint.
+    if (!filteredPaint || filteredPaint->getFilterQuality() != kLow_SkFilterQuality) {
         if (filteredPaint != &tmpPaint) {
             if (paint) {
                 tmpPaint = *paint;
-- 
2.7.4

